numba/targets/linalg.py

```diff
@@ -1110,3 +1110,285 @@ def qr_impl(a):
+    }
+
+    inv_type_table = {
+        np.dtype('float32'): types.float32,
+        np.dtype('float64'): types.float64,
+        np.dtype('complex64'): types.complex64,
+        np.dtype('complex128'): types.complex128
+    }
+
+    np_shared_dt = np.promote_types(type_table[a.dtype], type_table[b.dtype])
+    nb_shared_dt = inv_type_table[np_shared_dt]
+
+    # convert typing floats to np floats for use in the impl
+    r_type = getattr(nb_shared_dt, "underlying_float", nb_shared_dt)
+    if r_type.bitwidth == 32:
+        real_dtype = np.float32
+    else:
+        real_dtype = np.float64
+
+    # the lapack wrapper signature
+    numba_ez_gelsd_sig = types.intc(
+        types.char,  # kind
+        types.intp,  # m
+        types.intp,  # n
+        types.intp,  # nrhs
+        types.CPointer(nb_shared_dt),  # a
+        types.intp,  # lda
+        types.CPointer(nb_shared_dt),  # b
+        types.intp,  # ldb
+        types.CPointer(r_type),  # S
+        types.CPointer(r_type),  # rcond
+        types.CPointer(types.intc)  # rank
+    )
+
+    # the lapack wrapper function
+    numba_ez_gelsd = types.ExternalFunction("numba_ez_gelsd",
+                                            numba_ez_gelsd_sig)
+
+    kind = ord(get_blas_kind(nb_shared_dt, "lstsq"))
+
+    # The following functions select specialisations based on
+    # information around 'b', a lot of this effort is required
+    # as 'b' can be either 1D or 2D, and then there are
+    # some optimisations available depending on real or complex
+    # space.
+
+    # get a specialisation for computing the number of RHS
+    b_nrhs = _get_compute_nrhs(b)
+
+    # get a specialised residual computation based on the dtype
+    compute_res = _get_res_impl(nb_shared_dt, real_dtype, b)
+
+    # b copy function
+    b_copy_in = _get_copy_in_b_impl(b)
+
+    # return blob function
+    b_ret = _get_compute_return_impl(b)
+
+    # check system commutes function
+    check_commutes = _get_check_lstsq_commutes_impl(a, b)
+
+    def lstsq_impl(a, b, rcond=-1):
+        n = a.shape[-1]
+        m = a.shape[-2]
+        nrhs = b_nrhs(b)
+
+        # check the systems have no inf or NaN
+        _check_finite_matrix(a)
+        _check_finite_matrix(b)
+
+        # check the systems commute
+        check_commutes(a, b)
+
+        minmn = min(m, n)
+        maxmn = max(m, n)
+
+        # a is destroyed on exit, copy it
+        acpy = a.astype(np_shared_dt)
+        if a_F_layout:
+            acpy = np.copy(acpy)
+        else:
+            acpy = np.asfortranarray(acpy)
+
+        # b is overwritten on exit with the solution, copy allocate
+        bcpy = np.empty((nrhs, maxmn), dtype=np_shared_dt).T
+        # specialised copy in due to b being 1 or 2D
+        b_copy_in(bcpy, b, nrhs)
+
+        # Allocate returns
+        s = np.empty(minmn, dtype=real_dtype)
+        rcond_ptr = np.empty(1, dtype=real_dtype)
+        rank_ptr = np.empty(1, dtype=np.int32)
```

Comment:
Instead of allocating this, perhaps the rank can be returned by `numba_ez_gelsd` on success? (the rank is always >= 0, we currently only return 0 or negative)